db/seeds.rb

```diff
@@ -60,34 +61,84 @@
   setting.value = ModelConfig.generation.default.last&.id
 end
 
-# DEV
+# PROVISIONING
 #
-if Rails.env == "development"
-  ModelServer.find_or_create_by!(name: "localhost") do |ms|
-    ms.url = model_endpoint
-    ms.provider = "ollama"
-  end
-
-  ModelConfig.find_or_create_by!(name: "mistral:instruct", model_server: ModelServer.first) do |mc|
-    mc.model = "mistral:instruct"
-    mc.temperature = 0.1
-  end
-
-  ModelConfig.find_or_create_by!(name: "gemma:7b", model_server: ModelServer.first) do |mc|
-    mc.model = "gemma:7b"
-    mc.temperature = 0.2
-  end
-
-  ModelConfig.find_or_create_by!(name: "all-minilm", model_server: ModelServer.first) do |mc|
-    mc.model = "all-minilm"
-    mc.embedding = true
-  end
+devel_model_servers = [
+  {
+    name: "localhost",
+    provider: "ollama",
+  }
+]
+
+devel_model_configs = [
+  {
+    name: "mistral:instruct",
+    model_server: "localhost",
+    model: "mistral:instruct",
+    temperature: 0.1,
+  },
+  {
+    name: "gemma:7b",
+    model_server: "localhost",
+    model: "gemma:7b",
+    temperature: 0.2,
+  },
+  {
+    name: "all-minilm",
+    model_server: "localhost",
+    model: "all-minilm",
+    embedding: true,
+  },
+  {
+    name: "nomic-embed-text",
+    model_server: "localhost",
+    model: "nomic-embed-text",
+    embedding: true,
+  }
+]
+
+provisioned_model_servers = JSON.parse(ENV.fetch("PROVISIONED_MODEL_SERVERS") {
+  Rails.env == "development" ? JSON.generate(devel_model_servers) : '[]'
+})
+
+provisioned_model_configs = JSON.parse(ENV.fetch("PROVISIONED_MODEL_CONFIGS") {
+  Rails.env == "development" ? JSON.generate(devel_model_configs) : '[]'
+})
+
+
+provisioned_model_servers_names = provisioned_model_servers.map { |ms| ms["name"] }
+provisioned_model_config_names = provisioned_model_configs.map { |mc| mc["name"] }
+
+ModelConfig.where(provisioned: true).where.not(name: provisioned_model_config_names).update(enabled: false, provisioned: false)
+ModelServer.where(provisioned: true).where.not(name: provisioned_model_servers_names).update(enabled: false, provisioned: false)
+
+provisioned_model_servers.each do |config|
+  puts "provisioning model server for `#{config["name"]}` ..."
+
+  ModelServer.find_or_create_by!(name: config["name"]).update(
```

Comment:
This is tricky, because it only works if all the attributes in the `update` are optional. Looks like they all are, but they shouldn't be!

I think it's okay for now, but I'll need to make all these fields mandatory (e.g. no point in allowing a ModelServer without a `url`) and do something fancier at that point.

It's more common to:

```ruby
ModelServer.find_or_create_by!(name: "...") do |model_server|
  model_server.url = "..."
  ...
end
```

but you want to update all the attributes whether the model was created or not, so that wouldn't work for you.

Reply:
Yeah, I forgot to ask about the former.

As for the latter, that was my dilemma for like half a day. I'm honestly baffled as to why `do |model_server|` can't be called in either scenario. If you think about it, by the time we reach that point the underlying actions are the same—update an existing record:

1. **Create**: We couldn't find the record; so we've started the process of created one. Set any mandatory fields; and then, we'll call `.save` for you.
2. **Find**: We found the record. We'll update those fields you set during creation and call `.save` for you.

And it could still return the record, as it already does. Anywho, nothing I can do there. That said, I guess the appropriate way to do this would be something like:

```ruby
record = ModelServer.find_or_create_by!(name: "...") do |model_server|
  model_server.url = "..."
  ...
end

record.update(url: "...", ...)
```

Which is.... fine, I guess lol

Reply:
Will leave this for now, as I don't think like the idea of repeating the `create_by` code in the form of an `update`. There's a better, more abstract way to do this, I'm sure.